File: javascript/039/main.js
/*
 * ProjectEuler 039
 * JavaScript / main.js
 * copyright (c) 2015 Susisu
 *
 * Node.js v5.0.0
 * `node main.js`
 */

"use strict";

function main() {
    // p = a + b + c <= 1000
    // a^2 + b^2 = c^2
    // 0 < a <= b < c
    // -> a <= 333, b < 1000 - a - b
    var maxP = 1000;
    var numSolutions = [];
    var i;
    for (i = 0; i <= maxP; i++) {
        numSolutions[i] = 0;
    }
    for (var a = 1; a <= 333; a++) {
        for (var b = a; b < maxP - a - b; b++) {
            var cSquare = a * a + b * b;
            var c = Math.round(Math.sqrt(cSquare));
            // Is c an integer?
            if (Math.pow(c, 2) === cSquare) {
                var p = a + b + c;
                if (p <= maxP) {
                    numSolutions[p]++;
                }
            }
        }
    }
    var maxNumSolutions = 0;
    var maxNumSolutionsP = 0;
    for (i = 0; i <= maxP; i++) {
        if (numSolutions[i] > maxNumSolutions) {
            maxNumSolutions  = numSolutions[i];
            maxNumSolutionsP = i;
        }
    }
    console.log(maxNumSolutionsP);
}

main();

File: ocaml/039/main.ml
(*
    ProjectEuler 039
    OCaml / main.ml
    copyright (c) 2016 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main main.ml`
*)

let max_p = 1000

let _ =
    let num_solutions = Array.make max_p 0 in
    let rec count = function
        | 334, _ -> ()
        | a, b   ->
            let x = a * a + b * b in
            let c = int_of_float @@ floor @@ sqrt @@ float_of_int x in
            let p = a + b + c in
            if p > max_p then count (a + 1, a + 1)
            else begin
                if c * c = x then Array.set num_solutions (p - 1) (Array.get num_solutions (p - 1) + 1);
                count (a, b + 1)
            end
    in
    let rec find p m mp =
        if p > max_p then mp
        else
            let n = Array.get num_solutions (p - 1) in
            if n > m then find (p + 1) n p
            else find (p + 1) m mp
    in
    count (1, 1);
    let p = find 1 0 0 in
    print_int p;
    print_newline ()
